rclcpp/include/rclcpp/strategies/allocator_memory_strategy.hpp

```diff
@@ -379,10 +379,10 @@ class AllocatorMemoryStrategy : public memory_strategy::MemoryStrategy
 
   VectorRebind<const rcl_guard_condition_t *> guard_conditions_;
 
-  VectorRebind<const rcl_subscription_t *> subscription_handles_;
-  VectorRebind<const rcl_service_t *> service_handles_;
-  VectorRebind<const rcl_client_t *> client_handles_;
-  VectorRebind<const rcl_timer_t *> timer_handles_;
+  VectorRebind<std::pair<const rcl_subscription_t *, SubscriptionBase::SharedPtr>> sub_pair_ptrs_;
+  VectorRebind<rclcpp::ServiceBase::SharedPtr> service_ptrs_;
+  VectorRebind<rclcpp::ClientBase::SharedPtr> client_ptrs_;
+  VectorRebind<rclcpp::TimerBase::SharedPtr> timer_ptrs_;
```

Comment:
I think these should instead store `shared_ptr`'s to the `rcl_*` types. Here's my rationale why, consider this case (current behavior):

- user creates pub and sub which talk
- user spins
- asynchronously user reset's the sub pointer
- when a message is received for the deleted subscription:
  - a segfault occurs in `rcl_wait()` for using "finalized" pointer to a `rcl_subscription_t`

The existing behavior, described in https://github.com/ros2/rclcpp/issues/349, this resulted in segfault's when `rcl_wait` was still using an `rcl_*` object, because that object had had it's `_fini()` function called on it from the destructor of one of our C++ classes.

With the implementation in this pr it goes like this:

- user creates pub and sub which talk
- user spins
- asynchronously user reset's the sub pointer
- when a message is received for the deleted subscription:
  - a callback is called for the subscription the user has already deleted

With the proposal from #349, i.e. using `shared_ptr` to the `rcl_*` types, it would go like this (I think):

- user creates pub and sub which talk
- user spins
- asynchronously user reset's the sub pointer
- when a message is received for the deleted subscription:
  - the message is "dropped" because the `weak_ptr` to the subscription the executor has fails to materialize into a `shared_ptr` because it no longer exists